<p><img alt="caee logo" src="https://drive.google.com/uc?export=view&id=1bLjob5TsJBD3tWzNKzi6sbseyXsm6eb-" align="left" hspace="10px" vspace="0px"></p>
<p><img alt="iel logo" src="https://drive.google.com/uc?export=view&id=1uoCP8K9MyHTmU5nnpdlZoprhO8Bm5Yq5" align="right" hspace="10px" vspace="0px"></p>
<h2 align='center'>Homework 0: Python Introduction</h2>

ARE 346N Building Environmental Systems


Professor Dr. Zoltan Nagy


# Instructions

Solve each of the following problems and show your work. Make sure to label each answer with the same label as the problem.

You will be graded on both the correct method and solution. Partial credit is awarded for use of the correct equation or method, and the correct answer with appropriate units. 

This assignment includes *eight* problems.

| PROBLEM | POINTS |
| --- | --- |
| Problem 1 | 10 pts |
| Problem 2 | 15 pts |
| Problem 3 | 10 pts |
| Problem 4 | 15 pts |
| Problem 5 | 15 pts |
| Problem 6 | 15 pts |
| Problem 7 | 20 pts |
| **TOTAL** | **100 pts** |

# **Problems**

## Problem 1 - Creating a short program using `if` statements

---

Write a Python code  allows user to convert cooling/heating capacity from btu/hr to kW. Your input can be any variable related to energy such as cooling/heating capacities. 
$$
P(kW) = \left ( \frac{P[Btu/hr]}{3412.142} \right)
$$


In [ ]:
cooling_capacity = input("Input the energy you like to convert? (e.g., 100btu/hr, 1000kW etc.) : ") 
input_convert = cooling_capacity[-2:]

if input_convert == "hr":
  result = int(int(cooling_capacity[:-6])/ 3412.142)
  output_convert = "kW"
elif input_convert == "kW":
  result = int(int(cooling_capacity[:-2])* 3412.142)
  output_convert = "Btu/hr"
else:
  print("Input proper convention.")
  quit()
print("The energy in", output_convert, "is", result)


Input the energy you like to convert? (e.g., 100btu/hr, 1000kW etc.) : 100kW
The energy in Btu/hr is 341214


## Problem 2 - Creating a short program using `if` statements

---

Consider the following assessment criteria which assess the thermal comfort.

| Thermal comfort   | Fanger Seven Point Scale   ($-3<$ and $< 3$)      |
| ---------------- | ---------------------- |
| Cold        | $= -3$            |
| Cool        | $= -2$  |
| Slightly Cool            | $= -1$   |
| Neutral | $= 0$    |
| Slightly Warm     | $= +1$  |
| Warm     | $= +2$  |
| Hot    | $= +3$              |

Write a program that using `if`, `elif`, and `else` that asks for user input and prints the appropriate thermal comfort. Print an error message if the input is greater than +3 or less than -3.


In [ ]:
thermal_comfort = float(input("How do you feel? Please provide a number that shows your comfort criteria: "))
if (thermal_comfort == -3):
    print('If I understand correct, you are feeling cold')
elif (thermal_comfort == -2):  
    print('If I understand correct, you are feeling cool')
elif (thermal_comfort == -1):  
    print('If I understand correct, you are feeling slightly Cool')
elif (thermal_comfort == -0):  
    print('If I understand correct, you are feeling slightly Neutral')   
elif (thermal_comfort == 1):  
   print('If I understand correct, you are feeling slightly Slightly Warm')
elif (thermal_comfort == 2):  
    print('If I understand correct, you are feeling warm') 
elif (thermal_comfort == 3):  
    print('If I understand correct, you are feeling hot')
else:
    print("Out of the scale, There is an error in your input")


How do you feel? Please provide a number that shows your comfort criteria: 5
Out of the scale, There is an error in your input


## Problem 3 - Complex Order of Operations

For the polynomial 
\begin{align}
f(x, y, z) &= (x + y+ z )^{4} 
\\
&=  x^4 + y^4 +z^4 + 4x^{3}y + 4xy^{3}+ 4y^{3}x + 4yz^{3} +4z^{3}x+4x^{3}z+6x^{2}y^{2}+6y^{2}z^{2}+6z^{2}x^{2}+12x^{2}yz+12xy^{2}z+12xyz^{2}
\end{align}

compute $f$ using: (i) the compact form $(x + y+ z)^{4}$; and (ii) the expanded form for:



$x = 3$ and $y = 4.1$ and $z = 5.6$

and compare the number of significant digits for which the answers are the same.

Store the answer for the compact version using the variable `f0`, and using the variable `f1` for the expanded version.

For case (b), compare the computed and analytical solutions and consider the relative error.
Which approach would you recommend for computing this expression?

In [ ]:
x = 3
y = 4.1
z = 5.6


f0 = (x+y+z)**4
f1 = (x**4+ y**4+z**4+ 4*(x**3)*y+4*x*(y**3)+4*(y**3)*x+4*y*(z**3)+4*(z**3)*x+4*(x**3)*z+6*(x**2)*(y**2)+6*(y**2)*(z**2)+6*(z**2)*(x**2)+12*(x**2)*y*z+12*x*(y**2)*z+12*x*y*(z**2))


In [ ]:
f0

26014.464099999994

In [ ]:
f1

25297.685699999995

In [ ]:
error = (f0-f1)/f0
error

0.027553071908177396

`f0` = 26014.464 and `f1` = 25297.685699999995

error = 2%

**Since the error percentage is small we should think applying the first f0 approach because it is more time effective.**

## Problem 4 - Basic Function

Write a Python **function** to find the sum of the numbers listed in different columns.

|Variables | X | Y |
| ---   | --- | --- |
| X1, Y1| 10 | 4 |
|X2, Y2| 15 | 16 |
| X3, Y3| 18 | 33|
|X4, Y4| 55 | 47 |
| X5, Y5| 22 | 88 |
| **TOTAL** | ???|??? |



In [ ]:
def sum(numbers):
    total = 0
    for x in numbers:
        total += x
    return total

In [ ]:
print(sum((10, 15, 18, 55, 22)))

120


In [ ]:
print(sum((4, 16, 33, 47, 88)))

188


## Problem 5 - Writing Function with `if statement`

Write a Python function to check whether indoor temperature is in a given range below. If it is below 65 oF, turn on the heater. If the temperature is above 75 oF, turn on the cooler.

$$
65F <Tindoor < 75F
$$

---

Test 2 different temperatures above and below given temperature values to discuss your function.


In [ ]:
def test_indoorTemperature(T_indoor):
    if T_indoor in range(65,75):
        print( " %s is in the range"%str(T_indoor))
    elif   65> T_indoor:
        print("The temperature is below 65F, please turn on the heater")
    elif 75<T_indoor:
        print("The temperature is above 75F, please turn the cooler on." )
    else:
        print("enter reasonable Tindoor")

In [ ]:
test_indoorTemperature(100)

The temperature is above 75F, please turn the cooler on.


In [ ]:
test_indoorTemperature(10)

The temperature is below 65F, please turn on the heater


## Problem 6 - Function for different conditions

Write a function to calculate the condition of mixed air for given Tindoor, Toutdoor and indoor and outdoor air mixing ratio.
$$
T_(mixed) = \left( {Tindoor * returnair + Toutdoor * OutdoorAir} \right)
$$
Write a function named mixed air temperatur that computes for given the variables (4 different variables).





In [ ]:
# YOUR FUNCTION HERE
def mixed_air(T_indoor, Return_Air, T_outdoor, Outdoor_Air):
  return T_indoor*Return_Air+T_outdoor*Outdoor_Air

In [ ]:
T_indoor1, Return_Air1, T_outdoor1, Outdoor_Air1 = 75, 0.30, 85, 0.70   
T_indoor2, Return_Air2, T_outdoor2, Outdoor_Air2 = 63, 0.20, 40, 0.80
T_indoor3, Return_Air3, T_outdoor3, Outdoor_Air3 = 80, 0.50, 100, 0.50
T_indoor4, Return_Air4, T_outdoor4, Outdoor_Air4 = 70, 0.90, 110, 0.10

print(mixed_air(T_indoor1, Return_Air1, T_outdoor1, Outdoor_Air1))  # Expect 82F
print(mixed_air(T_indoor2, Return_Air2, T_outdoor2, Outdoor_Air2))  # Expect 44.6F
print(mixed_air(T_indoor3, Return_Air3, T_outdoor3, Outdoor_Air3))  # Expect 90F
print(mixed_air(T_indoor4, Return_Air4, T_outdoor4, Outdoor_Air4))  # Expect 74F

82.0
44.6
90.0
74.0


## Problem 7 - Creating a Dataframe




### Problem 7 -Part a

Please `create a dataset` as shown in the figure below, except the Pass column.

|Student Name | Year Accepted | Year of Graduation | Final Grade |
| ---   | --- | --- | --- |
| Tina| 2012 | 2016 | A |
|Sangeetha| 2017 | 2021 | A-|
| Joseph| 2016 | 2020| B- |
|Jake| 2012 | 2016 | B+ |
|Rachel| 2016 | 2020 | A- |
|Sep| 2018 | 2020 | A |
|June| 2016 | 2020 | A |
|Jose| 2018 | 2021 | A |


In [ ]:
#YOUR CODE HERE

In [ ]:
# Import modules
import pandas as pd

# Create an example dataframe
data = {'Name': ['Tina', 'Sangeetha', 'Joseph', 'Jake', 'Rachel', 'Sep', 'June', 'Jose'], 
        'Year': [2012, 2017, 2016, 2012, 2016, 2018, 2016, 2018], 
        'Graduation': [2016, 2021, 2020, 2016, 2020, 2020, 2020, 2021],
        'Grade': ['A','A-','B-', 'B+', 'A-','A','A','A']}
df = pd.DataFrame(data)
df

Name  Year  Graduation Grade
0       Tina  2012        2016     A
1  Sangeetha  2017        2021    A-
2     Joseph  2016        2020    B-
3       Jake  2012        2016    B+
4     Rachel  2016        2020    A-
5        Sep  2018        2020     A
6       June  2016        2020     A
7       Jose  2018        2021     A

### Problem 7 - Part b

Please add a column to the dataframe created previously. If the grade of the student less than "A-", student will fail.

In [ ]:
import numpy as np
# create a list of our conditions
conditions = [
    (df['Grade'] == 'A'),
    (df['Grade'] == 'A-'),
    (df['Grade'] == 'B-'),
    (df['Grade'] == 'B'),
    (df['Grade'] == 'B+')]

# create a list of the values we want to assign for each condition
Fail_Pass = ['Pass','Pass', 'Fail','Fail','Fail']

# create a new column and use np.select to assign values to it using our lists as arguments
df['Fail/Pass'] = np.select(conditions, Fail_Pass)

# display updated DataFrame
df


Name  Year  Graduation Grade Fail/Pass
0       Tina  2012        2016     A      Pass
1  Sangeetha  2017        2021    A-      Pass
2     Joseph  2016        2020    B-      Fail
3       Jake  2012        2016    B+      Fail
4     Rachel  2016        2020    A-      Pass
5        Sep  2018        2020     A      Pass
6       June  2016        2020     A      Pass
7       Jose  2018        2021     A      Pass

In [ ]:
import pandas as pd

def Fail_Pass(a):
  if row['Grade'] == 'A':
    return 'Pass'
  elif row['Grade'] == 'A-':
    return 'Pass'
  elif row['Grade'] == 'B':
    return 'Fail'
  elif row['Grade'] == 'B+':
    return 'Fail'
  elif row['Grade'] == 'B-':
    return 'Fail'
  else:
    return 'Undefined'

df['Fail/Pass'] = df.apply(Fail_Pass, axis=1)
df


Name  Year  Graduation Grade Fail/Pass
0       Tina  2012        2016     A      Pass
1  Sangeetha  2017        2021    A-      Pass
2     Joseph  2016        2020    B-      Fail
3       Jake  2012        2016    B+      Fail
4     Rachel  2016        2020    A-      Pass
5        Sep  2018        2020     A      Pass
6       June  2016        2020     A      Pass
7       Jose  2018        2021     A      Pass